### PR TITLE
fix(peft): sync shared expert LoRA weights across EP once per step

### PR DESCRIPTION
**Pull request by HollowMan6**
# What does this PR do ?

Shared expert LoRA weights (the default shared adapter and the shared factor of `experts_shared_outer_loras=True`) are replicated across expert-parallel ranks, but Megatron-Core's expert DDP only reduces them over expert-DP. This PR moves the missing cross-EP gradient sum from a per-layer backward hook to a single coalesced all-reduce over the EP group after `finalize_model_grads`, and makes it work with `gradient_accumulation_fusion`.

Two problems with the old hooks:

- **Silently broken with fused weight-gradient accumulation.** With `gradient_accumulation_fusion=True` (the default in most PEFT recipes) Megatron-Core writes the weight gradient into `main_grad` and hands autograd a dummy tensor, so the hook all-reduced a dummy and the shared weights drifted apart across EP ranks. Reproduced on a tiny MoE with EP=2: after 6 steps the shared weights differ by up to 0.09 across EP ranks on `main`, and by 0 with this PR.
- **Expensive for shared-outer.** Its hook ran a blocking all-reduce over the whole pipeline stage (TP x CP x DP) twice per MoE layer per microbatch, acting as a per-layer barrier across data-parallel replicas. That is why long-sequence steps looked communication-bound. It is now one all-reduce per step over the EP group only (measured with NCCL logging: 2 x layers x microbatches collectives over the stage -> 1 over EP).

# Changelog

- `peft/utils.py`: add `mark_expert_parallel_replicated`, `allreduce_expert_parallel_replicated_grads`, `enable_expert_parallel_grad_sync_in_finalize` and the `finalize_model_grads_with_expert_adapter_sync` wrapper; both shared expert adapters tag their shared weights and drop the per-layer hooks once the finalize path is active. A fallback hook keeps external training loops (fusion off) working unchanged; with fusion on it warns once on rank 0.
- `training/setup.py`: install the finalize wrapper as `finalize_model_grads_func` (no-op for models without shared expert adapters).
- Shared-outer adapter: broadcast and reduce over the EP group instead of the whole stage; accepts `pg_collection` like the other adapters.
- Tests: mock unit tests, a two-rank CPU (gloo) test, new cases in the two-rank NCCL test (finalize path and shared-outer), setup wiring test.
- Docs: `docs/training/peft.md` (and Fern mirror) describe the mechanism and the one-line wiring for training loops that call `finalize_model_grads` themselves.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci) in the Contributing doc for how to trigger the CI.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

# Additional Information

- Verified on 2x/4x H20 (EP=2, expert-DP=1 and 2, fusion on and off, both adapter layouts): loss curves match `main` with fusion off to 1e-4, shared weights are bitwise identical across EP after training.
- External training loops that set `finalize_model_grads_func` themselves (NeMo-RL, verl, Miles) keep working as before with fusion off. To also cover fusion on, replace Megatron-Core's `finalize_model_grads` with `finalize_model_grads_with_expert_adapter_sync` there.

### CODE REVIEW

File: src/megatron/bridge/peft/utils.py [hunk 1]
```diff
@@ -940,6 +944,175 @@ def all2all_hp2sp(input_: torch.Tensor, tensor_parallel_group: object | None = N
...
+    if not (torch.distributed.is_available() and torch.distributed.is_initialized()):
+        return
+    if ep_group is None:
+        ep_group = _get_process_group(_get_pg_collection(required_pgs=["ep"]), "ep")
+    if ep_group is None or _process_group_size(ep_group) <= 1:
+        return
+
+    grads_by_dtype: dict[tuple[torch.dtype, torch.device], list[torch.Tensor]] = {}
+    for param in _expert_parallel_replicated_params(model):
+        hook_was_active = _remove_fallback_hook(param)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply src/megatron/bridge/peft/utils.py]: Good catch on the ordering, fixed now: the tagged parameters are collected first and the function returns before any distributed or process-group lookup when there are none, so models without shared expert adapters never touch the MPU state.

One correction on the failure mode: the `ep` entry in `ProcessGroupCollection.use_mpu_process_groups` is `partial(get_expert_model_parallel_group, check_initialized=False)`, so the fallback lookup returns `None` (and the function returns) rather than raising when EP is not initialized. The reorder is still worthwhile to skip the per-step lookup for non-PEFT models.

File: tests/unit_tests/peft/test_utils.py [hunk 2]
```diff
@@ -1121,6 +1126,62 @@ def test_parallel_linear_adapter_grouped_expert_shared_adapter_syncs_cpu_init_ac
...
+        mock_linear_in, mock_linear_out = self._make_shared_expert_adapter(mock_config, mock_col_linear)
+        for weight in (mock_linear_in.weight, mock_linear_out.weight):
+            # MCore's fused path sets this before returning its dummy weight gradient.
+            weight.grad_added_to_main_grad = True
+
+        with (
+            patch("torch.distributed.all_reduce") as mock_all_reduce,
+            pytest.warns(UserWarning, match="main_grad"),
+        ):
+            (mock_linear_in.weight.sum() + mock_linear_out.weight.sum()).backward()
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply tests/unit_tests/peft/test_utils.py:1164]: Pinned `RANK=0` for the duration of that test now, so `get_rank_safe()` no longer emits its "Could not determine rank" warning.

For the record the assertion was not flaky: `pytest.warns(match=...)` passes as long as one recorded warning matches and the repo does not escalate warnings to errors, and the test was green on a 2-GPU box. The pin just makes the captured warnings deterministic.

File: src/megatron/bridge/training/setup.py [hunk 3]
```diff
@@ -643,7 +647,12 @@ def _update_model_config_funcs(
         if len(model) == 1:
             model_config.param_sync_func = model_config.param_sync_func[0]
     if optimizer is not None:
-        model_config.finalize_model_grads_func = partial(finalize_model_grads, pg_collection=pg_collection)
+        # Shared expert adapters are replicated across EP: sum their gradients once per step after the DP sync
+        # instead of once per layer and microbatch (a no-op for models without such adapters).
+        enable_expert_parallel_grad_sync_in_finalize(model)
```
Comment (yaoyu-33) [inline src/megatron/bridge/training/setup.py]: if peft enabled do this? if peft only

Comment (HollowMan6) [reply src/megatron/bridge/training/setup.py]: Done, `_update_model_config_funcs` now takes `peft_enabled` (passed as `cfg.peft is not None`) and only installs `finalize_model_grads_with_expert_adapter_sync` and drops the fallback hooks for PEFT runs. Non-PEFT runs keep Megatron-Core's `finalize_model_grads` unchanged. Covered by the updated `test_update_model_config_funcs_installs_expert_adapter_grad_sync_for_peft_only`.

**Review by yaoyu-33 — APPROVED** [review]

**Comment by huvunvidia** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/6009/commits/90437e4b1383e85bea7bf7e27a3b9ecbb13346c0

**Comment by huvunvidia** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/6009/commits/f9216eaf48e970e6d978f22ccf71cdc998d05e3d

**Comment by huvunvidia** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/6009/commits/7f7942c441ccee4f01366ad024d73457d158da72
